### PR TITLE
feat(auction): allow orders to automatically exit after partial fill

### LINKED ISSUE #7088: allow auction Bids to specify "exit on partial satisfaction"
## What is the Problem Being Solved?

Individual bids in the Dutch Auction (#7000) should be able to specify that their seat in the auction should be exited after successfully bidding on assets, regardless of whether they were entirely consumed. This provides support for arbitrageurs whose goal in the auction is to immediately sell the proceeds at the current market price in order to take advantage of the discount they got in the purchase.

This is in contrast with bidders whose goal is to accumulate assets at a discount, and who therefore want to let their bid ride and be applied to future rounds. (Those bidders can have an `onDemand` exit condition and pull assets out occasionally.)

## Description of the Design

This would be another field in offerArgs when exercising the invitation.

## Security Considerations

This fits within the Zoe security architecture, and is well within the ability of a contract to carry out.

## Scaling Considerations

No impact.

## Test Plan

Nothing outlandish.

### PR DESCRIPTION
**Pull request by Chris-Hibbert**
closes: #7088

## Description

Allow Auction bids to add an offerArgs parameter that says the seat should be exited if the offer is partially filled.

### Security Considerations

None

### Scaling Considerations

None.

### Documentation Considerations

One more detail for the bidding UI and CLI.

### Testing Considerations

Well tested.

### CODE REVIEW

File: packages/inter-protocol/src/auction/auctionBook.js [hunk 1]
```diff
@@ -71,6 +73,7 @@ export const makeBidSpecShape = (currencyBrand, collateralBrand) => {
   return M.splitRecord(
     { want: collateralAmountShape },
     {
+      exitAfterBuy: true,
```
Comment (turadg) [inline packages/inter-protocol/src/auction/auctionBook.js]: why is `exitAfterBuy: false` invalid?

if it truly is then make it the same in the TS type,
```
  exitAfterBuy?: true,
```

File: packages/inter-protocol/src/auction/auctionBook.js [hunk 2]
```diff
@@ -300,8 +305,15 @@ export const prepareAuctionBook = (baggage, zcf) => {
...
          *  @param {boolean} trySettle
+         *  @param {boolean} [exitAfterBuy]
          */
-        acceptScaledBidOffer(seat, bidScaling, want, trySettle) {
+        acceptScaledBidOffer(
+          seat,
+          bidScaling,
+          want,
+          trySettle,
+          exitAfterBuy = false,
```
Comment (turadg) [inline packages/inter-protocol/src/auction/auctionBook.js]: with two booleans now, please make it an opts object. it's too easy to get the order wrong and nothing would complain

File: packages/inter-protocol/src/auction/auctionBook.js [hunk 3]
```diff
@@ -462,52 +477,40 @@ export const prepareAuctionBook = (baggage, zcf) => {
...
           const { give } = seat.getProposal();
+          const { currencyAmountShape } = this.state;
           mustMatch(
             give.Currency,
             currencyAmountShape,
             'give must include "Currency"',
           );
 
           const { helper } = this.facets;
+          const exitAfterBuy = bidSpec.exitAfterBuy;
```
Comment (turadg) [inline packages/inter-protocol/src/auction/auctionBook.js]: nit
```suggestion
          const { exitAfterBuy }  = bidSpec;
```

File: packages/inter-protocol/src/auction/auctionBook.js (hunk 1 above)
Comment (Chris-Hibbert) [reply packages/inter-protocol/src/auction/auctionBook.js]: I think you're right that the mistake is here. `exitAfterBuy` is a boolean that defaults to `false`.

File: packages/inter-protocol/src/auction/auctionBook.js (hunk 2 above)
Comment (Chris-Hibbert) [reply packages/inter-protocol/src/auction/auctionBook.js]: done

File: packages/inter-protocol/src/auction/auctionBook.js (hunk 3 above)
Comment (Chris-Hibbert) [reply packages/inter-protocol/src/auction/auctionBook.js]: done

**Comment by Chris-Hibbert** [comment]
@turadg I added `automerge:squash` so this can go in tonight if you approve. If you want to approve with requested changes, please remove the label before approving.

**Comment by erights** [comment]
From the PR comment:

> Allow Auction bids to add an offerArgs parameter that says the seat should be exited if the offer is partially exited.

Did you mean "partially filled"?

**Review by turadg — APPROVED** [review]
Thanks for the automerge label comment. This does look mergable to me, modulo resolving conflicts

**Comment by Chris-Hibbert** [comment]
> Did you mean "partially filled"?

Yes. Corrected.

**Pull request merged**